From b9cd6e8965a8702a91bb1b1ba5569fe02ec93c52 Mon Sep 17 00:00:00 2001
From: Terry Zhou <bjzhou@marvell.com>
Date: Thu, 23 Jun 2016 09:31:59 +0300
Subject: [PATCH 227/239] fix: a3700: add cci-400 initialization in BL3

- The cci-400 initialization used to be done in SPL, but it was
  missing when moving the SPL code to ATF. The snooping function
  is not enabled on the coherent fabric, which leads into IO
  stop working.
- The coherent bus is supposed to be initialized in ATF stage.
  Since we didn't enable USE_CCI option in ATF, it's risky to
  enable the cci_init code in ATF. Also, cci_enable_snoop_dvm_reqs()
  cannot fully meet the cci configurations on S3 port.
- As a temporary solution, we moved the original spl code to
  Armada3700's dedicated bl31_plat_arch_setup() since it can only
  be done in BL3 stage.

Change-Id: Ie8e6795bd242318ec12ff63649b3e1961c43c115
Signed-off-by: Terry Zhou <bjzhou@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/30686
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Tested-by: Wilson Ding <dingwei@marvell.com>
---
 plat/marvell/a3700/a3700_z/plat_bl31_setup.c | 17 +++++++++++++++++
 plat/marvell/a3700/a3700_z/plat_def.h        |  4 ++++
 2 files changed, 21 insertions(+)

diff --git a/plat/marvell/a3700/a3700_z/plat_bl31_setup.c b/plat/marvell/a3700/a3700_z/plat_bl31_setup.c
index 16408cc..d08068a 100644
--- a/plat/marvell/a3700/a3700_z/plat_bl31_setup.c
+++ b/plat/marvell/a3700/a3700_z/plat_bl31_setup.c
@@ -114,6 +114,20 @@ static void marvell_boot_mode_switch(void)
 		marvell_boot_mode_set(BOOT_MODE_UART);
 }
 
+void misc_init_cci400(void)
+{
+	/*
+	   CCI-400 enable snoop and dvm on S3 port.
+	   For details see the <CoreLink CCI-400 Cache Coherent Interconnect> document.
+	   bit[0] - Enable issuing of snoop requests from this slave interface.
+	   bit[1] - Enable issuing of DVM message requests from this slave interface
+	   bit[29:2] - Reserved
+	   bit[30] - Slave interface supports snoops
+	   bit[31] - Slave interface supports DVM messages
+	 */
+	mmio_write_32(MVEBU_CCI_S3_SNOOP_CTRL_REG, 0xC0000003);
+}
+
 /* This function overruns the same function in marvell_bl31_setup.c */
 void bl31_plat_arch_setup(void)
 {
@@ -133,4 +147,7 @@ void bl31_plat_arch_setup(void)
 
 	/* Pass DRAM size value so that u-boot could get it later */
 	pass_dram_sys_info();
+
+	/* initialiaze the cci400 */
+	misc_init_cci400();
 }
diff --git a/plat/marvell/a3700/a3700_z/plat_def.h b/plat/marvell/a3700/a3700_z/plat_def.h
index 9f4d804..85c16e5 100644
--- a/plat/marvell/a3700/a3700_z/plat_def.h
+++ b/plat/marvell/a3700/a3700_z/plat_def.h
@@ -83,4 +83,8 @@
 /* I2C register base address*/
 #define MVEBU_I2C_REGS_BASE			(MVEBU_REGS_BASE + 0x11000)
 
+/* CCI-400 */
+#define MVEBU_CCI_BASE			(MVEBU_REGS_BASE + 0x8000000)
+#define MVEBU_CCI_S3_SNOOP_CTRL_REG	(MVEBU_CCI_BASE + 0x4000)
+
 #endif /* __MVEBU_DEF_H__ */
-- 
1.9.1

